### ISSUE TITLE
The FIFO stress test hangs

### ISSUE THREAD

**Issue by AntonioND** [blocksds/sdk#49]
The following test hangs eventually, after hundreds of thousands of words have been sent from one CPU to the other one.

https://github.com/blocksds/sdk/tree/094e15db6ead2ea1c6ab06957288ce0a422f1da9/tests/fifo_stress_test

This may have been present in the code before forking libnds, though.

Also, sometimes it doesn't hang but one of the messages received by one of the handlers is incorrect, which may mean that libnds is just sending the message to the wrong channel.

**Comment by AntonioND** [comment]
After this commit https://github.com/blocksds/libnds/commit/34528657d628baf63d8c0a2eed0890bec23d0a0c it doesn't hang anymore, but I'm going to keep this issue open because I want to also test data messages and addresses.

**Comment by AntonioND** [comment]
After this commit https://github.com/blocksds/sdk/commit/efb1d4fb00a9e7939fe4f17c56c1b5f3703317b3 the test sets up an address, data and immediate handler for each of the user channels. Then, it starts the stress test. It seems to be reliable enough. When it isn't, it crashes after a second or so. I've run this for a couple of minutes without crashes, and it can exit to the loader, so I'm going to say that this is working reliably now.

**Comment by AntonioND** [comment]
Huh, maybe I spoke too soon.

**Comment by AntonioND** [comment]
If I fully disable nested interrupts in the FIFO handler, it becomes reliable to the point that it never crashes and it can always exit to loader.

Before doing that, after a second or two I can no longer exit to the loader because of memory corruption.

However, the interesting thing is that nested interrupts only happen when there is a callback to handle messages sent to that CPU. In the ARM9 I'm not using that for anything. There are two explanations:

- ARM7 libnds system messages (such as input) are received in the ARM9 in callbacks. It's possible that nested interrupts inside their callbacks corrupt the stack.

- The ARM7 code of that test sets up callbacks for all messages. Maybe it's the ARM7 the one that corrupts the main RAM. However, this is unlikely, because that code is much smaller than the ARM9 part of the test.

**Comment by AntonioND** [comment]
The multithreading example was also crashing, and I tracked down the issue to be the variable length array in ``fifoGetDatamsg()``.

https://github.com/blocksds/libnds/commit/53a1d8e6bc4b3924854dc1b61015cb427fe8d74b

However, the FIFO stress test isn't working reliably yet (there is another VLA in a FIFO function, maybe that's the reason.

**Comment by AntonioND** [comment]
Mystery solved! Eventually I managed to reproduce the test consistently in no$gba.

This bug was caused by stack corruption. ``fifo_buffer_enqueue_block()`` ``SP=0x2FF3BF0`` writes to the wrong address while simply pushing the callee-saved registers to the stack. It has been called by ``fifoInternalRecvInterrupt()`` ``SP=0x2ff3c38`` caused by a FIFO interrupt.

That interrupt has happened during ``fifoGetDatamsg()``, ``SP=0x2FF3C40`` which was called by ``systemMsgHandler()``, ``SP=0x2FF3C98`` which was called by ``fifoInternalRecvInterrupt()``, caused by a FIFO Interrupt.

The interrupt has happened during ``swiWaitForIRQ()`` ``SP=0x2FF3CD8``, which was called by ``cothread_scheduler_start()``. This happens because all threads are idle waiting for FIFO messages or the VBL, which makes sense, the CPU meter of no$gba isn't even half full.

Now, what's the interesting part? I only reserve 0x100 (256) bytes for the stack of the scheduler!

The map file says:

```
0x02ff3d00  __sp_usr = (__sp_irq - 0x100) 
```

This is reserved for the scheduler, and the crt0 defines the start of the stack of the main thread ``0x100`` bytes below that point, at around ``0x02ff3c00``.

Basically, this nested interrupt situation was taking the stack pointer from ``0x2FF3CD8``, almost at the top of the scheduler stack, to ``0x2FF3BF0``, after the end of the scheduler stack.
